tests/Progress/ProgressUpdateRepositoryTest.php

```diff
@@ -0,0 +1,183 @@
+        $this->activityInstance = factory(ActivityInstance::class)->create(['activity_id' => $this->activity->id]);
+
+        $this->hashesTableName = (new ProgressHashes)->getTable();
+    }
+
+    /** @test */
+    public function saving_progress_saves_a_new_hash_correctly()
+    {
+        // Create a new Snapshot Instance and trigger Update to Activity Instance:
+        (new Snapshot($this->repository))->ofUpdateToActivityInstance($this->activityInstance, 'caller');
+
+        // Ensure that there is a new hash in the DB:
+        $this->assertDatabaseHas($this->hashesTableName, ['item_key' => 'caller_' . $this->activityInstance->id]);
+    }
+
+    /**
+     * @test
+     * @dataProvider progressChangeProvider
+     */
+    public function saving_progress_updates_a_hash_when_a_progress_is_changed(\Closure $getData)
+    {
+        [$firstProgress, $changedProgress] = $getData();
+
+        $this->repository->saveChanges($this->activityInstance->id, 'some_caller',  $firstProgress);
+
+        $this->assertTrue(
+            $this->repository->hasChanged($this->activityInstance->id, 'some_caller', $changedProgress)
+        );
+    }
+
+    public function progressChangeProvider(): array
+    {
+        return [
+            'Return two set of Activities with Modules' => [
+                function () {
+                    return [
+                        $this->createProgress(),
+                        $this->createProgress(function (Progress $progress): Progress {
+                            $progress->setActivityId(factory(Activity::class)->create()->id);
+                            $progress->setActivityInstanceId(factory(ActivityInstance::class)->create()->id);
+                            return $progress;
+                        })
+                    ];
+                }
+            ],
+            'Return two set of Activities with different percentages' => [
+                function () {
+                    return [
+                        $this->createProgress(),
+                        $this->createProgress(function (Progress $progress): Progress {
+                            $progress->setActivityId(factory(Activity::class)->create()->id);
+                            $progress->setPercentage(50);
+                            return $progress;
+                        })
+                    ];
+                }
+            ],
+            'Return two set of Activities with different complete statuses' => [
+                function () {
+                    return [
+                        $this->createProgress(),
+                        $this->createProgress(function (Progress $progress): Progress {
+                            $progress->setActivityId(factory(Activity::class)->create()->id);
+                            $progress->isComplete(false);
+                            return $progress;
+                        })
+                    ];
+                }
+            ]
```

Comment:
This should return a provider for every condition mentioned in `ProgressUpdateRepository::generateActivityString`, so we make sure we're always triggering progress updates when any of the attributes change. Should just be a copy and paste job to add a few more!